<p align="center">
  <img src="https://github.com/hrayjones/foreseen/blob/main/foreseen.png?raw=true" width="300">
</p>

<p align="center">
***** UNDER DEVELOPMENT *****
</p>

## Foreseen: An automated pipeline for 4C-seq experimental design, with or without genetic variants of interest

4C-seq is a powerful method for profiling the 3D genome at specific genomic regions of interest. Depending on the design, a 4C-seq experiment can be used to identify chromatin interactions in a genotype-specific fashion, which has clear utility in resolving the effect of genotype on gene regulatory mechanisms. However, designing 4C-seq is complicated and time-consuming. Moreover, a manual design process does not necessarily explore all possibilities for experimental design. The foreseen pipeline automates the design process, taking as input a genomic locus and running through all possible combinations of valid restriction enzymes and primers to suggest the most optimal experimenal parameters for the 4C-seq experiment.

### Requirements
- perl

### Installation
*installation instructions*

### Basic usage
*basic usage instructions*

#### The modes are as follows:
1) __Basic__ (classic 4C-seq; no genotype-specificity required)
2) __Allele-specific__ (4C-seq that is focused on a specific allele of a single nucleotide polymorphism [SNP] of interest)
3) __Allele-aware__ (4C-seq that yields separable sequencing results for both alleles of a SNP of interest in a heterozygous sample)

## How does Foreseen work?


